homeassistant/components/mcp/__init__.py

```diff
@@ -0,0 +1,108 @@
+"""The Model Context Protocol integration."""
+
+from __future__ import annotations
+
+from dataclasses import dataclass
+
+import httpx
+from mcp import types
+from voluptuous_openapi import convert_to_voluptuous
+
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import HomeAssistantError
+from homeassistant.helpers import llm
+from homeassistant.util.json import JsonObjectType
+
+from .const import DOMAIN
+from .coordinator import ModelContextProtocolCoordinator
+from .types import ModelContextProtocolConfigEntry
+
+__all__ = [
+    "DOMAIN",
+    "async_setup_entry",
+    "async_unload_entry",
+]
+
+API_PROMPT = "The following tools are available from a remote server named {name}."
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ModelContextProtocolConfigEntry
+) -> bool:
+    """Set up Model Context Protocol from a config entry."""
+    coordinator = ModelContextProtocolCoordinator(hass, entry)
+    await coordinator.async_config_entry_first_refresh()
+
+    unsub = llm.async_register_api(
+        hass,
+        ModelContextProtocolAPI(
+            hass=hass,
+            id=f"{DOMAIN}-{entry.entry_id}",
+            name=entry.title,
+            coordinator=coordinator,
+        ),
+    )
+    entry.async_on_unload(unsub)
+
+    entry.runtime_data = coordinator
+    entry.async_on_unload(coordinator.close)
+
+    return True
+
+
+async def async_unload_entry(
+    hass: HomeAssistant, entry: ModelContextProtocolConfigEntry
+) -> bool:
+    """Unload a config entry."""
+    return True
+
+
+class ModelContextProtocolTool(llm.Tool):
+    """A Tool exposed over the Model Context Protocol."""
+
+    def __init__(
+        self,
+        tool: types.Tool,
+        coordinator: ModelContextProtocolCoordinator,
+    ) -> None:
+        """Initialize the tool."""
+        self.name = tool.name
+        self.description = tool.description
+        self.parameters = convert_to_voluptuous(tool.inputSchema)
```

Comment:
When this raises, should we somewhere log the input schema that brought it down? Currently it will bubble all the way up to the conversation entity I think?

Reply:
Moved this so that it fails coordinator / integration setup instead.